### PR TITLE
Add read optimized CLUSTER command.

### PR DESCRIPTION
**Pull request by JLockerman**
Adds a version of `CLUSTER` which

1. Updates indicies via the same copy-to-new-location-then-swap as the data tables.
2. Does not block reads until the final swap.

Also adds infrastructure to use postgres's `pg_isolation_regress` tool to test multi-transaction behavior.

### CODE REVIEW

File: src/recluster.c [hunk 1]
```diff
@@ -0,0 +1,1175 @@
...
+			case HEAPTUPLE_RECENTLY_DEAD:
+				tups_recently_dead += 1;
+				/* fall through */
+			case HEAPTUPLE_LIVE:
+				/* Live or recently dead, must copy it */
+				isdead = false;
+				break;
+			case HEAPTUPLE_INSERT_IN_PROGRESS:
+
+				/*
```
Comment (cevian) [inline src/recluster.c:532]: Given the comment here, I would throw an error for this case since this cannot be a system catalog.

File: src/recluster.c [hunk 2]
```diff
@@ -0,0 +1,1175 @@
...
+				 */
+				if (!TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetXmin(tuple->t_data)))
+					elog(WARNING, "concurrent insert in progress within table \"%s\"",
+						 RelationGetRelationName(OldHeap));
+				/* treat as live */
+				isdead = false;
+				break;
+			case HEAPTUPLE_DELETE_IN_PROGRESS:
+
+				/*
```
Comment (cevian) [inline src/recluster.c:545]: Also throw error here.

File: src/recluster.c [hunk 3]
```diff
@@ -0,0 +1,1175 @@
...
+	{
+		Oid			old_index_oid = lfirst_oid(old_index_cell);
+		Oid			new_index_oid = lfirst_oid(new_index_cell);
+
+		swap_relation_files(old_index_oid, new_index_oid,
+		/* TODO */
+							swap_toast_by_content, true,
+							frozenXid, cutoffMulti);
+	}
+	/* TODO assert same length? */
```
Comment (cevian) [inline src/recluster.c]: Yes please add this assert. Also would it be safer to pass the old and new indexoids lists along instead of recomputing the old list. That seems a bit dangerous?

File: test/isolation/expected/isolation_nop.out [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Parsed test spec with 1 sessions
```
Comment (cevian) [inline test/isolation/expected/isolation_nop.out:1]: What is this for?

File: test/isolation/specs/timescale_cluster.spec [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+teardown { DROP TABLE ts_cluster_test; }
+
+session "s1"
+setup		{ BEGIN; }
+step "s1a"	{ SELECT * FROM ts_cluster_test; }
+step "s1b"	{ INSERT INTO ts_cluster_test VALUES ('2017-01-22T09:00:01', 19.5, 3); }
+step "s1c"	{ COMMIT; }
+
+session "s2"
+step "s2"	{ CLUSTER ts_cluster_test USING ts_cluster_test_time_idx; }
```
Comment (cevian) [inline test/isolation/specs/timescale_cluster.spec:19]: This does all possible permutations?

File: test/sql/cluster.sql [hunk 6]
```diff
@@ -40,3 +40,18 @@ CLUSTER VERBOSE cluster_test using cluster_test_time_location_idx;
...
 WHERE indisclustered = true;
+
+-- we should start read_optimized
+SELECT setting FROM pg_settings WHERE name = 'timescaledb.cluster_method';
+
+-- Set guc to run basic cluster
+UPDATE pg_settings SET setting = 'native' WHERE name = 'timescaledb.cluster_method';
+
+-- and old cluster should be run
+CLUSTER VERBOSE cluster_test using cluster_test_time_location_idx;
```
Comment (cevian) [inline test/sql/cluster.sql]: using should be in CAPS

File: test/sql/cluster.sql [hunk 7]
```diff
@@ -40,3 +40,18 @@ CLUSTER VERBOSE cluster_test using cluster_test_time_location_idx;
...
+
+-- Set guc back
+UPDATE pg_settings SET setting = 'read_optimized' WHERE name = 'timescaledb.cluster_method';
+
+-- and new cluster should be run
+CLUSTER VERBOSE cluster_test using cluster_test_time_location_idx;
```
Comment (cevian) [inline test/sql/cluster.sql]: can we add some tests:
1) Clustering on chunks (not hypertables)
2) Clustering on tables with toast fields
3) Clustering on tables with multiple indexes

File: src/process_utility.c [hunk 8]
```diff
@@ -1104,13 +1106,19 @@ process_cluster_start(Node *parsetree, ProcessUtilityContext context)
...
 	/* If this is a re-cluster on all tables, there is nothing we need to do */
+	/* TODO should we take charge and run our cluster on our tables? */
 	if (NULL == stmt->relation)
 		return false;
 
 	hcache = hypertable_cache_pin();
 	ht = hypertable_cache_get_entry_rv(hcache, stmt->relation);
 
-	if (NULL != ht)
+	if (NULL == ht)
```
Comment (cevian) [inline src/process_utility.c:1187]: Why this change? Also why is the below still indented?

File: src/guc.c [hunk 9]
```diff
@@ -91,6 +98,17 @@ _guc_init(void)
 							NULL,
 							assign_max_cached_chunks_per_hypertable_hook,
 							NULL);
+
+	DefineCustomEnumVariable("timescaledb.cluster_method", "Enable cluster with reduced locking.",
```
Comment (cevian) [inline src/guc.c]: I think "Enable" here doesn't make sense as this is an enum. Maybe "Choose..."

File: src/chunk_index.c [hunk 10]
```diff
@@ -925,6 +925,55 @@ chunk_index_mark_clustered(Oid chunkrelid, Oid indexrelid)
...
+	Relation	src_chunk_rel;
+	Relation	dest_chunk_rel;
+	List	   *index_oids;
+	Oid			constraint_oid;
+	ListCell   *index_elem;
+	List	   *new_index_oids = NIL;
+	Chunk	   *src_chunk;
+
+	/* TODO lock order? */
+	src_chunk_rel = relation_open(src_chunkrelid, AccessShareLock);
```
Comment (cevian) [inline src/chunk_index.c]: would be nice to keep all the opens together. i.e. move line 949 up. The order of open and close should be reversed. also why mix relation_open and heap open?

File: src/chunk_index.c [hunk 11]
```diff
@@ -925,6 +925,55 @@ chunk_index_mark_clustered(Oid chunkrelid, Oid indexrelid)
 	heap_close(rel, AccessShareLock);
 }
 
+/*
+ * Create versions of every index over src_chunkrelid over chunkrelid.
+ * Returns the relids of the new indexes created.
+ * New indexes are in the same order as RelationGetIndexList.
+ */
+List *
+chunk_index_duplicate(Oid src_chunkrelid, Oid dest_chunkrelid)
```
Comment (cevian) [inline src/chunk_index.c]: Can we refactor this and chunk_index_clone to use common code?

File: src/recluster.c [hunk 12]
```diff
@@ -0,0 +1,1175 @@
+/*-------------------------------------------------------------------------
+ *
```
Comment (cevian) [inline src/recluster.c:2]: should we do a cluster_import.c/h akin to planner_import where we put pg functions that are completely unmodified? This lets us know right away which parts have not been touched. Which of these functions would qualify if we did that. To be clear: If there are functions we did modify to get rid of code inside we should keep those modified in this file.

File: src/recluster.c [hunk 13]
```diff
@@ -0,0 +1,1175 @@
...
+	 * to cluster a not-previously-clustered index.
+	 */
+	if (recheck)
+	{
+		HeapTuple	tuple;
+		Form_pg_index indexForm;
+
+		/* Check that the user still owns the relation */
+		if (!pg_class_ownercheck(tableOid, GetUserId()))
+		{
```
Comment (cevian) [inline src/recluster.c]: Should this be an error?

File: src/recluster.c [hunk 14]
```diff
@@ -0,0 +1,1175 @@
...
+			relation_close(OldHeap, ExclusiveLock);
+			return;
+		}
+
+		/*
```
Comment (cevian) [inline src/recluster.c]: Let's throw an error for any kind of temp table. "Other" or not.

File: src/recluster.c [hunk 15]
```diff
@@ -0,0 +1,1175 @@
...
+				ReleaseSysCache(tuple);
+				relation_close(OldHeap, ExclusiveLock);
+				return;
+			}
+			ReleaseSysCache(tuple);
+		}
+	}
+
+	/*
+	 * We allow VACUUM FULL, but not CLUSTER, on shared catalogs.  CLUSTER
```
Comment (cevian) [inline src/recluster.c]: Change comment. We don't allow recluster on shared catalogs period.

File: src/recluster.c [hunk 16]
```diff
@@ -0,0 +1,1175 @@
...
+	 * indisclustered in the current database, leading to unexpected behavior
+	 * if CLUSTER were later invoked in another database.
+	 */
+	if (OldHeap->rd_rel->relisshared)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot recluster a shared catalog")));
+
+	/*
+	 * Don't process temp tables of other backends ... their local buffer
```
Comment (cevian) [inline src/recluster.c]: Is this a duplicated check (see comment above)?

File: src/recluster.c [hunk 17]
```diff
@@ -0,0 +1,1175 @@
...
+	 * Also check for active uses of the relation in the current transaction,
+	 * including open scans and pending AFTER trigger events.
+	 */
+	/* TODO is this check needed/valid? */
+	CheckTableNotInUse(OldHeap, "CLUSTER");
+
+	/* Check heap and index are valid to cluster on */
+	check_index_is_clusterable(OldHeap, indexOid, recheck, ExclusiveLock);
+
+	/*
```
Comment (cevian) [inline src/recluster.c]: Throw an error on any materialized view.

File: src/recluster.c [hunk 18]
```diff
@@ -0,0 +1,1175 @@
...
+	 */
+	if (OldHeap->rd_rel->relkind == RELKIND_MATVIEW &&
+		!RelationIsPopulated(OldHeap))
+	{
+		relation_close(OldHeap, ExclusiveLock);
+		return;
+	}
+
+	/*
+	 * All predicate locks on the tuples or pages are about to be made
```
Comment (cevian) [inline src/recluster.c]: Can you look more into this. What exactly are predicate locks?

File: src/recluster.c (hunk 12 above)
Comment (cevian) [reply src/recluster.c:2]: Just to be clear I don't know if that's a good idea. Asking

File: src/recluster.c [hunk 19]
```diff
@@ -0,0 +1,1175 @@
...
+						get_namespace_name(RelationGetNamespace(OldHeap)),
+						RelationGetRelationName(OldHeap),
+						RelationGetRelationName(OldIndex))));
+	else if (tuplesort != NULL)
+		ereport(elevel,
+				(errmsg("reclustering \"%s.%s\" using sequential scan and sort",
+						get_namespace_name(RelationGetNamespace(OldHeap)),
+						RelationGetRelationName(OldHeap))));
+	else
+		ereport(elevel,
```
Comment (cevian) [inline src/recluster.c]: Shouldn't this be an error for us?

File: test/sql/cluster.sql (hunk 7 above)
Comment (cevian) [reply test/sql/cluster.sql]: Also let's do both sequential and index scans on the tables after each cluster.

File: test/pg_isolation_regress.sh [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+        for t in ${EXE_DIR}/isolation/specs/*.spec; do
+            t=${t##${EXE_DIR}/isolation/specs/}
+            t=${t%.spec}
+            TESTS="${TESTS} $t"
+        done
+    else
+        PG_ISOLATION_REGRESS_OPTS="${PG_ISOLATION_REGRESS_OPTS} --schedule=${ISOLATION_TEST_SCHEDULE}"
+    fi
+else
+    FILTER=${TESTS}
```
Comment (cevian) [inline test/pg_isolation_regress.sh:29]: Add some comments about why

File: test/pg_regress.sh [hunk 21]
```diff
@@ -17,6 +17,16 @@ if [[ -z ${TESTS} ]]; then
     else
         PG_REGRESS_OPTS="${PG_REGRESS_OPTS} --schedule=${TEST_SCHEDULE}"
     fi
+else
+    FILTER=${TESTS}
```
Comment (cevian) [inline test/pg_regress.sh:29]: add comments

File: src/recluster.c [hunk 22]
```diff
@@ -0,0 +1,1175 @@
...
+	 * expectation that the other process will timeout and abort first.
+	 * Currently we set `deadlock_timeout` to 1 hour, as this should be longer
+	 * than any other normal process, while still allowing the system to make
+	 * progress in the event of a real deadlock. As this is the last lock we
+	 * grab, and the setting is local to our transaction we do not bother
+	 * changing the guc back.
+	 */
+	/* TODO is this the right timeout value? */
+	int			config_change = set_config_option("deadlock_timeout",
+												  "101000",
```
Comment (cevian) [inline src/recluster.c]: Please make this a #define at top of file.

**Review by cevian — CHANGES_REQUESTED** [review]
Very good job Josh. This is a great first PR. Some comments below. Note that these are not dictums but suggestions so feel free to respond with answers/rebuttals/etc. I am often wrong :)

File: src/guc.c (hunk 9 above)
Comment (JLockerman) [reply src/guc.c]: Going to rewrite that, now that it's an enum instead of a bool the text doesn't make sense.

File: src/process_utility.c (hunk 8 above)
Comment (JLockerman) [reply src/process_utility.c:1187]: Now that we're returning `true` in the case where we perform `CLUSTER` there wasn't any benifit from having a unified return at the end of the function instead of returning early when we encounter an unexpected state. Additionally, the late return made it harder for me to figure out what this function was doing when I started reading it.
I left the block and indent because changing it would move a very large amount of variables to the top of the function (I thought I'd seen this done elsewhere in the postgres codebase, though I cannot seem to find it ATM).

File: src/recluster.c (hunk 12 above)
Comment (JLockerman) [reply src/recluster.c:2]: First I'll have to check and see how much unmodified code is left; the garbage collection we did touched most of it.

File: test/isolation/expected/isolation_nop.out (hunk 4 above)
Comment (JLockerman) [reply test/isolation/expected/isolation_nop.out:1]: That's to test that filtering and command discovery work for our `pg_isolation_regress` wrapper. If it fails it means the test infrastructure is broken.

File: test/isolation/specs/timescale_cluster.spec (hunk 5 above)
Comment (JLockerman) [reply test/isolation/specs/timescale_cluster.spec:19]: Yup, no permutations specified means it tries all of them.

File: src/chunk_index.c (hunk 11 above)
Comment (JLockerman) [reply src/chunk_index.c]: This is surprisingly difficult because the two functions have very different control flow. `chunk_index_duplicate` starts with a chunk uses that to get the hypertable and from that the indices, while `chunk_index_clone` goes index -> chunk -> hypertable.

File: src/chunk_index.c (hunk 10 above)
Comment (JLockerman) [reply src/chunk_index.c]: Not sure why I made one a `relation_open`, changed it to `heap` like the others.
I'd like to keep all the `open`s together, but I need o get `src_chunk` from `src_chunk_rel` before I know which hypertable to open, so I think there will have to be a little intermixing.

File: src/recluster.c (hunk 13 above)
Comment (JLockerman) [reply src/recluster.c]: I don't believe so: CLUSTER is deliberately does not hold a lock for its entire duration, so there can be a sequence of events like

```
CLUSTER starts.
CLUSTER commits the initial transaction.
change ownership.
CLUSTER starts a transaction for one of the chunks.
```
However, we still might want to try CLUSTERing the other chunks, or, if we're working on all tables, another table. That said, I think we might want to log a warning.

File: src/recluster.c (hunk 18 above)
Comment (JLockerman) [reply src/recluster.c]: (basing on [this](https://wiki.postgresql.org/wiki/Serializable#Predicate_Locking)) it's a fine grained lock (can lock individual tuples) which is used to guarantee serializable transactions don't read entries which were moved into its WHERE clause after the transaction's serialization point. I'm not touching this without reading a whole lot more of the postgres lock implementation.

File: src/recluster.c (hunk 3 above)
Comment (JLockerman) [reply src/recluster.c]: `RelationGetIndexList` guarantees that the list is sorted in order by OID, and since we we're locking out writes the set of indices cannot be changed, so it should be safe, but I agree it'll be nicer to just return both lists.

File: test/sql/cluster.sql (hunk 7 above)
Comment (JLockerman) [reply test/sql/cluster.sql]: We do 3 at lines 35-37. 1 is not yet implemented, it just uses the regular CLUSTER. I don't know how to do 2 and 4 yet.

Comment (cevian) [reply test/sql/cluster.sql]: Ahh 1 should be part of this PR for sure.

File: src/recluster.c (hunk 12 above)
Comment (JLockerman) [reply src/recluster.c:2]: It looks like none of the functions are untouched except `reform_and_rewerite_tuple`.

File: src/process_utility.c (hunk 8 above)
Comment (cevian) [reply src/process_utility.c:1187]: Two things. 1) we don't use such block. a style thing. 2) a unified return is still useful because of the cleanup that has to be done before the return

File: test/isolation/expected/isolation_nop.out (hunk 4 above)
Comment (cevian) [reply test/isolation/expected/isolation_nop.out:1]: wouldn't the second test fail in any case?

File: src/recluster.c (hunk 12 above)
Comment (cevian) [reply src/recluster.c:2]: Ok then nevermind. Let's just keep it all in here.

File: src/recluster.c (hunk 13 above)
Comment (cevian) [reply src/recluster.c]: My preference here would be to be conservative and throw an error in this scenario. But we can leave for now and let others chime in.

**Comment by JLockerman** [comment]
Reclustering now works when called on a chunk directly.

File: src/process_utility.c [hunk 23]
```diff
@@ -1191,12 +1197,105 @@ process_cluster_start(Node *parsetree, ProcessUtilityContext context)
...
+
+	/* If the relation wasn't a chunk let postgres's cluster handle it. */
+	if (NULL == chunk)
+		return false;
+
+	ht = hypertable_cache_get_entry(hcache, chunk->hypertable_relid);
+
+	recluster_permission_check(ht, is_top_level);
+
+	/* first check the hypertable for the index, also contains the NULL check */
```
Comment (cevian) [inline src/process_utility.c]: I actually think the NULL case should be handled differently. If we get NULL we should find_clustered_index on the chunk not the hypertable. A hypertable cluster will mark the chunk index as the clustered index in `chunk_index_mark_clustered(cim->chunkoid, cim->indexoid);` right?

File: src/process_utility.c [hunk 24]
```diff
@@ -1191,12 +1197,105 @@ process_cluster_start(Node *parsetree, ProcessUtilityContext context)
...
+		cim = chunk_index_get_by_indexrelid(chunk, index_relid);
+	}
+
+	Assert(cim != NULL);
+	Assert(cim->chunkoid == chunk_id);
+	recluster_chunk_by_index_map(cim, stmt->verbose);
+	return true;
+}
+
+/* get te hypertable index for an indexname or InvalidOid if none exists */
```
Comment (cevian) [inline src/process_utility.c]: spelling

File: src/recluster.c [hunk 25]
```diff
@@ -0,0 +1,1140 @@
...
+	if (OldHeap->rd_rel->reltoastrelid)
+		LockRelationOid(OldHeap->rd_rel->reltoastrelid, ExclusiveLock);
+
+	/*
+	 * We need to log the copied data in WAL iff WAL archiving/streaming is
+	 * enabled AND it's a WAL-logged rel.
+	 */
+	use_wal = XLogIsNeeded() && RelationNeedsWAL(NewHeap);
+
+	/* use_wal off requires smgr_targblock be initially invalid */
```
Comment (cevian) [inline src/recluster.c:359]: This comment is weird since use_wal never used in Assert below

File: src/recluster.c [hunk 26]
```diff
@@ -0,0 +1,1140 @@
...
+				isdead = false;
+				break;
+			case HEAPTUPLE_INSERT_IN_PROGRESS:
+
+				/*
+				 * Since we hold exclusive lock on the relation, the only way
+				 * to see this is if it was inserted earlier in our own
+				 * transaction.
+				 */
+				if (!TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetXmin(tuple->t_data)))
```
Comment (cevian) [inline src/recluster.c:537]: I don't think we need this if, just throw an error. I don't think we ever insert in same transaction as we recluster in.

File: src/recluster.c [hunk 27]
```diff
@@ -0,0 +1,1140 @@
...
+						 RelationGetRelationName(OldHeap));
+				/* treat as live */
+				isdead = false;
+				break;
+			case HEAPTUPLE_DELETE_IN_PROGRESS:
+
+				/*
+				 * Similar situation to INSERT_IN_PROGRESS case.
+				 */
+				if (!TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetUpdateXid(tuple->t_data)))
```
Comment (cevian) [inline src/recluster.c:548]: also this if.

File: src/recluster.c [hunk 28]
```diff
@@ -0,0 +1,1140 @@
...
+	 * run once per chunk, we do not want to abort it due to this deadlock. To
+	 * prevent abort we set our `deadlock_timeout` to a large value in the
+	 * expectation that the other process will timeout and abort first.
+	 * Currently we set `deadlock_timeout` to 1 hour, as this should be longer
+	 * than any other normal process, while still allowing the system to make
+	 * progress in the event of a real deadlock. As this is the last lock we
+	 * grab, and the setting is local to our transaction we do not bother
+	 * changing the guc back.
+	 */
+	int			config_change = set_config_option("deadlock_timeout",
```
Comment (cevian) [inline src/recluster.c:708]: Should we undo this change at any point?

**Review by cevian — CHANGES_REQUESTED** [review]

File: src/recluster.c (hunk 28 above)
Comment (JLockerman) [reply src/recluster.c:708]: We don't grab any locks for the remainder of the transaction, and change is transaction local, and setting it to the previous value would involve round tripping it through a string.

File: src/recluster.c (hunk 26 above)
Comment (JLockerman) [reply src/recluster.c:537]: can the user create a write/delete in the same command as a CLUSTER? Because when clustering 1 chunk we stay within the same transaction.

File: src/recluster.c (hunk 25 above)
Comment (JLockerman) [reply src/recluster.c:359]: I think we might want to have this available for replication need be, and it passes all the tests anyway.

File: src/recluster.c (hunk 26 above)
Comment (cevian) [reply src/recluster.c:537]: I don't think so. in recluster_permission_check we have `PreventTransactionChain(is_top_level, "CLUSTER");` which means that no other commands can be in the same txn, no?

File: test/sql/cluster.sql [hunk 29]
```diff
@@ -34,9 +34,35 @@ CLUSTER VERBOSE;
 -- Change the clustered index
```
Comment (cevian) [inline test/sql/cluster.sql:145]: Bringing up another comment about tests:
1) We need toast fields to test this. See `toast_test` table in size_utils test.
2) We need a test for the dropped toast-column test. (Forcing pSwapByContent = false)
3) We need to test both seq and index scans after each cluster. An example
```
BEGIN;
  SET LOCAL enable_seqscan='off';
  EXPLAIN (costs off) SELECT * FROM cluster_test;
  SELECT * FROM cluster_test;
COMMIT;
```
The first line turn off seq scans, forcing an indexscan (to turn of indexscans the option is enable_indexscan). The second line prints the plan to check it's what we want. The third line executes the plan.
4) (optional) Right now we have some tests that do `using index scan` and some that do `using sequential scan and sort` But this seems to happen by accident. I think a test that explicitly does each would be better practice. I am not positive but I think adjusting `random_page_cost` option could force the plan from one to the other.

**Review by cevian — CHANGES_REQUESTED** [review]
